src/finalfusion/vocab/simple_vocab.py

```diff
@@ -0,0 +1,192 @@
+        self._words = words
+
+    @property
+    def words(self) -> list:
+        """
+        Get the list of known words
+
+        Returns
+        -------
+        words : List[str]
+            list of known words
+        """
+        return self._words
+
+    @property
+    def word_index(self) -> dict:
+        """
+        Get the index of known words
+
+        Returns
+        -------
+        dict : Dict[str, int]
+            index of known words
+        """
+        return self._index
+
+    @property
+    def idx_bound(self) -> int:
+        """
+        The exclusive upper bound of indices in this vocabulary.
+
+        Returns
+        -------
+        idx_bound : int
+           Exclusive upper bound of indices covered by the vocabulary.
+        """
+        return len(self.word_index)
+
+    def idx(self, item: str, default: Union[list, int, None] = None
+            ) -> Optional[Union[list, int]]:
+        """
+        Lookup the given query item.
+
+        This lookup does not raise an exception if the vocab can't produce indices.
+
+        Parameters
+        ----------
+        item : str
+            The query item.
+        default : Optional[Union[int, List[int]]]
+            Fall-back value to return if the vocab can't provide indices.
+
+        Returns
+        -------
+        index : Optional[Union[int, List[int]]]
+            * An integer if there is a single index for a known item.
+            * A list if the vocab can provide subword indices for a unknown item.
+            * The provided `default` item if the vocab can't provide indices.
+        """
+        return self.word_index.get(item, default)
+
+    def __getitem__(self, item: str) -> Union[list, int]:
+        return self.word_index[item]
+
+    def __contains__(self, item: Any) -> bool:
+        # usual case: checking whether a str is known
+        if isinstance(item, str):
+            return self.word_index.get(item) is not None
+        # e.g. allows checking whether one vocab is the superset of another
+        if hasattr(item, "__iter__"):
+            return all(w in self for w in item)
+        return False
+
+    def __iter__(self) -> Iterable[str]:
+        return iter(self.words)
+
+    def __len__(self) -> int:
+        return len(self.word_index)
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, type(self)):
+            return False
+        if self.words != other.words:
+            return False
+        if self.word_index != other.word_index:
+            return False
+        return True
+
+    @staticmethod
+    def read_chunk(file) -> 'SimpleVocab':
+        length = SimpleVocab._read_binary(file, "<Q")[0]
+        words, index = SimpleVocab._read_items(file, length)
+        return SimpleVocab(words, index)
+
+    def write_chunk(self, file):
+        SimpleVocab._write_binary(file, "<I", int(self.chunk_identifier()))
+        b_words = [bytes(word, "utf-8") for word in self.words]
```

Comment:
Perhaps there a better way than storing the full vocabulary in memory? In Rust UTF-8 encoding/decoding is very cheap, how much does it cost in Python? I guess I would mind less if saving took a few milliseconds more than suddenly using whatever amount of memory extra (especially for lower-memory applications).

Reply:
I changed this to encode on the fly and take the sum without storing the words in bytes. I agree, waiting a little longer is probably a nicer scenario than surprising someone with a blown up memory footprint.

Afaik, Python internally uses fixed width representations for characters to allow constant time indexing, so there must be some overhead for conversion, but I doubt that it's enough to actually matter.